### PR TITLE
Refactor functions/pmfs

### PR DESCRIPTION
**Pull request by adrian-lison**
This PR aims for a slight efficiency improvement in the computation of pmfs by roughly halfing the number of `normal_cdf` function evaluations:

- Assuming that the indices have unit distance (e.g. delay of 1,2,...,D days, without missing days in between), we can almost half the number of `normal_cdf` function evaluations by using shifted vectors (since the upper bound cdf value of one delay is the lower bound cdf value of the following delay).
- The above assumption fully satisfies the current implementation in `functions/pmfs`. If the goal is to support **leaving out** of individual days in the reporting delay distribution later on, this PR is not recommended. In contrast, if the goal is to support **binning** of individual days in the reporting delay distribution later on, this PR is also sensible aside from the potential efficiency improvements.

**Impact**
The changes do not influence posterior estimates. First testing on the vignette examples indicates slight efficiency gain, but more detailed testing using PR #41 should be done.

**Further adjustments**
- Normalization operation moved out of the for loop
- Slight adjustments in naming and position of declaration of `trunc_cdf` to achieve more consistency between the lognormal and the gamma cdf
- Change vertical bar in cdf functions to comma, in order to comply with the current stan doc. Vertical bar compiles with cmdstanr, but throws an error in stanr, making it impossible to use `stanr::expose_functions`

### CODE REVIEW

**Comment by adrian-lison** [comment]
Regarding the change from vertical bar in cdf functions to comma, it seems that this was once the convention (see https://groups.google.com/g/stan-users/c/SN3raHiPJZQ) but has been changed since and support for the comma will be removed in later versions of stan (see https://github.com/stan-dev/stan/pull/1898, and in particular https://github.com/stan-dev/stanc3/commit/d5890a8711acaef836d925c20f6ddb6fa0df607e)  It seems that the stan manual has not been updated. And since rstan is always far behind, they don't support the vertical bar notation yet.

This means that we will have to go with the vertical bar notation in the long run, and if rstan is too slow, sacrifice the ability to run `expose_functions` on the corresponding model code. Overall, it may thus not be worth it to switch to the comma notation now and revert commit #67a0ac0a71ab3145ca26c0752651f4852a2219ba.

**Review by seabbs — APPROVED** [review]
This is really great Adrian. Like the best things simple but clever. I think the generalisability sacrifice here is totally worth doing. One possible extension to this would be pre-calculating the indices in `calculate_pmfs` in R and passing in as those should be fixed I think.

Obviously the lack of formal regression and unit tests more broadly is not ideal now we are formalising this package but from running locally and visual inspection I am happy this does what it says on the tin. 


I'm very left a few comments that it would be good to resolve before merging this in.  

Generally, with PRs I try to have the author add a news line linking to the PR and flagging the primary author as a way of tracking dev (though this can now be done using GitHub releases not maybe required anymore) + updating the dev version number. Development up to now has just been me mostly so a bit of a mess so it would be good to formalise what we want to do along these lines.

**Review by seabbs — APPROVED** [review]

File: inst/stan/functions/pmfs.stan [hunk 1]
```diff
@@ -37,12 +40,15 @@ vector discretised_gamma_pmf(int[] y, real mu, real sigma, int max_val) {
...
-  for (i in 1:n){
-    pmf[i] = (gamma_cdf(y[i] + 1 | alpha, beta) - 
-                gamma_cdf(y[i] |  alpha, beta)) /
-    trunc_pmf;
+  real trunc_cdf = gamma_cdf(max_val + 1, alpha, beta);
+  vector[n] upper_cdf;
+  for (i in 1:n) {
+    upper_cdf[i] = gamma_cdf(y[i] + 1, alpha, beta);
   }
+  pmf[1] = upper_cdf[1] - gamma_cdf(y[1], alpha, beta);
```
[the remark this answers is not in the document]
Comment (adrian-lison) [reply inst/stan/functions/pmfs.stan:48]: Yes, I came across this when trying to test the function. If you think this is desirable we should keep the comma style for now and wait until it gets removed in stan 2.32 I think

File: inst/stan/functions/pmfs.stan [hunk 2]
```diff
@@ -7,23 +7,26 @@ vector discretised_lognormal_pmf(int[] y, real mu, real sigma, int max_val) {
   real c_mu = fmax(small, mu);
   vector[n] adj_y = to_vector(y) + small;
   vector[n] upper_y = (log(adj_y + 1) - c_mu) / c_sigma;
-  vector[n] lower_y = (log(adj_y) - c_mu) / c_sigma;
+  real lowest_y = (log(adj_y[1]) - c_mu) / c_sigma;
```
[the remark this answers is not in the document]
Comment (adrian-lison) [reply inst/stan/functions/pmfs.stan:10]: Yes, I thad thought about this as well... I am fine with going for the more optimal, specific version if you also think that we are likely to be happy with it. Also, we can easily switch back if we change our minds.

**Comment by adrian-lison** [comment]
Thanks @seabbs for the review. :) Some more remarks/ideas:

- If you think that we will very likely always have this hard-coded sequence of delays from 0:n, then we can get rid of the indices altogether I think and implement this directly in the loop of the corresponding pmf.
- There is a more memory-efficient version of my code which does not use `upper_cdf` as a separate vector, but instead only uses `pmf` in two separate steps, overwriting the values. It is however less interpretable and people would have to handle it with care because the order of computation matters. I didn't implement it so far because I thought that the local memory requirement is not a big factor here.
- Yep, I can add a news item and update the dev number, but what do you mean by flagging the author precisely?

File: inst/stan/functions/pmfs.stan (hunk 1 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply inst/stan/functions/pmfs.stan:48]: Yes good point. I don't think it makes a difference in run times but obiviously having to be stuck on older stan is not ideal. I assume that we also need to stick with the old way of specifying arrays (int[, ] vs array[, ] int) as well to maintain this functionality? How annoying.

File: inst/stan/functions/pmfs.stan (hunk 2 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply inst/stan/functions/pmfs.stan:10]: Agree. Sounds good.

**Comment by seabbs** [comment]
I had a bit of time so I made the additions we discussed above. I also further simplified so that:

- The gamma and lognormal functions share more code
- There is no lower bound used as it is 0 and evaluates to 0
- No upper bound is calculated as this is always the maximum delay
- Switched to using the stan parameterisation of the `lognormal_cdf` as now more stable
- Switched the gamma distribution to use the shape and scale parameterisation. This now works whereas before it was very unstable. 

I also changed the argument name used to change the distribution from `dist` to `distribution`. Tested in comparison to master using the README example only. I'd be keen on a review of this as quite a few code changes.

**Comment by seabbs** [comment]
I'm pretty happy with this so keen to merge. @adrian-lison no pressure to review as I realise we have speeding into a holder season but let me know if happy to let this one into `main` :)  (also yes copy pasta :)

**Review by adrian-lison — COMMENTED** [review]
Nicely done, looks all correct and very clean now. Will commit some minor typo fix and an error handling statement. Ready to merge!

**Pull request merged**